include/rmlab/auxiliary/Filesystem.cpp

```diff
@@ -0,0 +1,174 @@
+#   include <windows.h>
+#else
+#   include <cstring>
+#   include <sys/stat.h>
+#   include <sys/types.h>
+#   include <dirent.h>
+#endif
+
+#include <sstream>
+#include <stdexcept>
+#include <system_error>
+
+
+namespace rmlab
+{
+namespace auxiliary
+{
+bool
+directory_exists(std::string const& path)
+{
+#ifdef _WIN32
+    DWORD attributes = GetFileAttributes(path.c_str());
+
+    return (attributes != INVALID_FILE_ATTRIBUTES &&
+            (attributes & FILE_ATTRIBUTE_DIRECTORY));
+#else
+    struct stat s;
+    return (0 == stat(path.c_str(), &s)) && S_ISDIR(s.st_mode);
+#endif
+}
+
+bool
+file_exists( std::string const& path )
+{
+#ifdef _WIN32
+    DWORD attributes = GetFileAttributes(path.c_str());
+
+    return (attributes != INVALID_FILE_ATTRIBUTES &&
+            !(attributes & FILE_ATTRIBUTE_DIRECTORY));
+#else
+    struct stat s;
+    return (0 == stat(path.c_str(), &s)) && S_ISREG(s.st_mode);
+#endif
+}
+
+std::vector< std::string >
+list_directory(std::string const& path )
+{
+    std::vector< std::string > ret;
+#ifdef _WIN32
+    std::string pattern(path);
+    pattern.append("\\*");
+    WIN32_FIND_DATA data;
+    HANDLE hFind = FindFirstFile(pattern.c_str(), &data);
+    if( hFind == INVALID_HANDLE_VALUE )
+        throw std::system_error(std::error_code(errno, std::system_category()));
+    do {
+        if( strcmp(data.cFileName, ".") && strcmp(data.cFileName, "..") )
+            ret.emplace_back(data.cFileName);
+    } while (FindNextFile(hFind, &data) != 0);
+    FindClose(hFind);
+#else
+    auto directory = opendir(path.c_str());
+    if( !directory )
+        throw std::system_error(std::error_code(errno, std::system_category()));
+    dirent* entry;
+    while ((entry = readdir(directory)) != nullptr)
+        if( strcmp(entry->d_name, ".") && strcmp(entry->d_name, "..") )
+            ret.emplace_back(entry->d_name);
+    closedir(directory);
+#endif
+    return ret;
+}
+
+bool
+create_directories( std::string const& path )
+{
+    if( directory_exists(path) )
+        return true;
+
+#ifdef _WIN32
+    auto mk = [](std::string const& p) -> bool { return CreateDirectory(p.c_str(), nullptr); };
+#else
+    mode_t mask = umask(0);
+    umask(mask);
+    auto mk = [mask](std::string const& p) -> bool { return (0 == mkdir(p.c_str(), 0777 & ~mask));};
+#endif
+    std::istringstream ss(path);
+    std::string token;
+
+    std::string partialPath;
+    // path starts with directory separator? keep also in partialPath
+    if( path.compare(0, 1, &directory_separator) == 0)
+        partialPath += directory_separator;
+    bool success = true;
+    while( std::getline( ss, token, directory_separator ) )
+    {
+        if( !token.empty() )
+            partialPath += token + directory_separator;
+        if( !directory_exists( partialPath ) )
+        {
+            bool partial_success = mk(partialPath);
+            if( !partial_success )
+                // did someone else just race us to create this dir?
```

Comment:
You can probably skip the race condition checks in your application (as I would assume there are no parallel writes).

Reply:
Yes, thought so for a minute and decided to keep it. Doesn't harm :)